File: app/src/main/java/com/example/p2invest/WelcomeActivity.java
package com.example.p2invest;

import android.annotation.SuppressLint;
import android.app.ActivityManager;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.AlphaAnimation;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

public class WelcomeActivity extends AppCompatActivity {

    private static final int TO_MAIN = 1;
    private TextView tv_welcome_version;
    private RelativeLayout rl_welcome;
    @SuppressLint("HandlerLeak")
    private Handler handler=new Handler(){
        @Override
        public void handleMessage(@NonNull Message msg) {
            super.handleMessage(msg);
            if (msg.what==TO_MAIN){
                finish();
                Intent intent = new Intent(WelcomeActivity.this,MainActivity.class);
                startActivity(intent);
            }
        }
    };
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_welcome);

        initView();

    }

    private void initView() {
        tv_welcome_version = (TextView) findViewById(R.id.tv_welcome_version);
        rl_welcome = (RelativeLayout) findViewById(R.id.rl_welcome);

        AlphaAnimation alphaAnimation=new AlphaAnimation(0,1);
        alphaAnimation.setDuration(2000);
        alphaAnimation.setInterpolator(new AccelerateInterpolator());
        rl_welcome.startAnimation(alphaAnimation);



        handler.sendEmptyMessageDelayed(TO_MAIN, 3000);
    }
}

File: app/src/main/java/com/example/p2pfiancial/activity/WelcomeActivity.kt
package com.example.p2pfiancial.activity

import android.app.Activity
import android.content.Context
import android.net.ConnectivityManager
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.Window
import android.view.WindowManager
import android.view.animation.AccelerateInterpolator
import android.view.animation.AlphaAnimation
import android.view.animation.Interpolator
import com.example.commen.ActivityInstanceManager
import com.example.p2pfiancial.R
import com.example.p2pfiancial.util.UIUtils
import kotlinx.android.synthetic.main.activity_welcome.*
import org.jetbrains.anko.startActivity

class WelcomeActivity : Activity() {
    private var startTime: Long = 0

    private var handler: Handler = object : Handler() {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                TO_MAIN -> {
                    finish()
                    startActivity<MainActivity>()
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //去掉窗口标题
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        //隐藏顶部的状态栏
        window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
        setContentView(R.layout.activity_welcome)


        //将当前的activity添加到ActivityManager中
        ActivityInstanceManager.addActivity(this)
        //提供启动动画
        setAnimation()

        //联网更新应用
        updateApkFile()
    }

    /**
     * 联网更新应用
     */
    private fun updateApkFile() {
        startTime = System.currentTimeMillis()

        //1. 判断手机是否可以联网
        val connect: Boolean = isConnect()

        if (!connect) { //没有网络
            UIUtils.toast(getString(R.string.app_activity_welcome_no_net), false)
            toMain()

        }else{ //有网络
            toMain()
        }
    }

    private fun toMain() {
        val currentTime = System.currentTimeMillis()
        var delayTime = 3000 - (currentTime - startTime)
        if (delayTime < 0) {
            delayTime = 0
        }

        handler.sendEmptyMessageDelayed(TO_MAIN, delayTime)
    }


    /**
     * 提供启动动画
     */
    private fun setAnimation() {
        val alphaAnimation = AlphaAnimation(0.0f, 1.0f) //0: 完全透明, 1:完全不透明
        alphaAnimation.duration = 3000
        alphaAnimation.interpolator = AccelerateInterpolator() as Interpolator? //设置动画的变化率

        //启动动画
        rl_welcome.startAnimation(alphaAnimation) //添加动画
    }

    //判断是否联网
    private fun isConnect(): Boolean {
        var connected = false
        val manager =
            getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = manager.activeNetworkInfo
        if (networkInfo != null) {
            connected = networkInfo.isConnected
        }

        return connected
    }

    companion object {
        const val TO_MAIN = 1
        const val DOWNLOAD_VERSION_SUCCESS = 2
        const val DOWNLOAD_APK_FAIL = 3
        const val DOWNLOAD_APK_SUCCESS = 4
    }

    private var oldTime:Long = 0
    override fun onBackPressed() {
        val currentTime = System.currentTimeMillis()
        if (currentTime - oldTime > 2000) {
            UIUtils.toast(getString(R.string.app_activity_welcome_back_pressed_toast), false)
        }else{
            finish()
        }
        oldTime = currentTime
    }
}
